packages/assets-controllers/src/MultichainTokensRatesController/MultichainTokensRatesController.ts

```diff
@@ -0,0 +1,387 @@
+   */
+  #isNonEvmAccount(account: InternalAccount): boolean {
+    return (
+      !isEvmAccountType(account.type) && account.metadata.snap !== undefined
+    );
+  }
+
+  /**
+   * Retrieves all multichain accounts from the AccountsController.
+   *
+   * @returns An array of internal accounts.
+   */
+  #listMultichainAccounts(): InternalAccount[] {
+    return this.messagingSystem.call(
+      'AccountsController:listMultichainAccounts',
+    );
+  }
+
+  /**
+   * Filters and returns non-EVM accounts that should have balances.
+   *
+   * @returns An array of non-EVM internal accounts.
+   */
+  #listAccounts(): InternalAccount[] {
+    const accounts = this.#listMultichainAccounts();
+    return accounts.filter((account) => this.#isNonEvmAccount(account));
+  }
+
+  /**
+   * Retrieves a non-EVM account by its ID.
+   *
+   * @param accountId - The account ID.
+   * @returns The corresponding internal account, or undefined if not found.
+   */
+  #getAccount(accountId: string): InternalAccount | undefined {
+    return this.#listAccounts().find((account) => account.id === accountId);
+  }
+
+  /**
+   * Handles the removal of an account by deleting its conversion rates.
+   *
+   * @param accountId - The ID of the removed account.
+   */
+  async #handleOnAccountRemoved(accountId: string): Promise<void> {
+    if (accountId in this.state.conversionRates) {
+      this.update((state: Draft<MultichainTokensRatesControllerState>) => {
+        delete state.conversionRates[accountId];
+      });
+    }
+  }
+
+  /**
+   * Updates the token conversion rates for a given account.
+   *
+   * This method acquires a mutex lock to ensure thread safety.
+   *
+   * @param accountId - The account ID for which to update rates.
+   * @returns A promise that resolves when the rates have been updated.
+   */
+  async updateTokensRates(accountId: string): Promise<void> {
+    const releaseLock = await this.#mutex.acquire();
+
+    return (async () => {
+      if (!this.isActive) {
+        return;
+      }
+
+      const account = this.#getAccount(accountId);
+
+      if (!account?.metadata.snap) {
+        return;
+      }
+
+      // Retrieve assets from the assets controller.
+      const assets = await this.#getAssetsList(
+        accountId,
+        account.metadata.snap.id as SnapId,
+      );
+
+      const accountType = account.type as KeyringAccountType;
+
+      const conversions = assets.map((asset) => ({
+        from: asset,
+        to: MAP_SWIFT_ISO4217[accountType],
+      }));
+
+      const accountRates = await this.#handleSnapRequest({
+        snapId: account.metadata.snap.id as SnapId,
+        handler: 'onAssetsConversion' as HandlerType,
+        conversions,
+      });
+
+      // Update the state with new conversion rates.
+      this.update((state: Draft<MultichainTokensRatesControllerState>) => {
+        state.conversionRates[accountId] = accountRates.conversionRates;
```

Comment:
Shouldn't we store the rates per assets and not per account ? This means we'll have duplicated values when a snap has multiple accounts

Reply:
this is because previously i was relying on the selectedAccount , but now this will change with the refacto i'm doing

Reply:
I think it would be easier if we aggregated all the tokens coming from a snap and store the rates with the token ID as the key. This would dedupe a lot.


```ts
type MultichainTokensRatesControllerState = {
  [tokenId: CaipTokenId]: TokenConversionRates
}
```

Reply:
yes yes , that's what i'm going to do now